dali/kernels/slice/slice_hwc2chw_normalize_gpu.cu

```diff
@@ -0,0 +1,531 @@
+ *    and cast it to float.
+ *   a. Unaligned prologue loop - reads the first chunk till we get to address that is aligned with
+ *      32 * 4.
+ *   b. Main loop - do as many aligned 4byte reads as possible
+ *   c. Epilogue loop - read the remaining values that were not possible to read as one 4byte read.
+ * 2. Synchronize
+ * 3. [Output loop] Each thread corresponds to a (Y, X) sequential offset into a plane, computes
+ *    the values for all the channels and writes them.
+ *   a. Optionally, mirroring is performed by inverting the X-coordinate in the output offset.
+ *   b. Padding the output channels is performed by filling additional planes with fill values.
+ *  @{
+ */
+
+/**
+ * @brief Hwc2Chw Normalize Mirror-x Pad-channel kernel
+ * This kernel does not support cropping the x coordinate, so the reads are fully linear.
+ */
+template <typename Out, typename In, bool enable_mirror, bool enable_pad, int kBlockSize,
+          int kStaticChannels>
+__global__ void Hwc2ChwNormalize(const Hwc2ChwSampleDesc<Out, In> *samples,
+                                 uint32_t num_samples) {
+  // TODO(klecki): generalize for wider input types
+  static_assert(std::is_same<In, uint8_t>::value, "Only uint8_t supported as input");
+
+  int sample_idx = FindSampleIdx(samples, num_samples);
+  const auto sample = samples[sample_idx];
+  int64_t start_x = (blockIdx.x - sample.first_block) * kBlockSize;
+  int64_t end_x = ::min(start_x + kBlockSize, sample.sample_size);
+  __shared__ float tile[kBlockSize + 32 * 4];
+
+  // Preload the norm values so they are accessed via registers and not from gmem via pointer.
+  float norm_mul[kStaticChannels], norm_add[kStaticChannels];
+
+  #pragma unroll kStaticChannels
+  for (int c = 0; c < kStaticChannels; c++) {
+    norm_mul[c] = sample.norm_mul[c];
+    norm_add[c] = sample.norm_add[c];
+  }
+
+  auto in_start = reinterpret_cast<std::uintptr_t>(sample.in + start_x);
+  auto aligned_in_start = align_up(in_start, 32 * 4);
+  auto bytes_skipped =
+      ::min(static_cast<int64_t>(aligned_in_start - in_start), end_x - start_x);
+
+  float *aligned_tile = tile + 32 * 4;
+  float *prologue_tile = aligned_tile - bytes_skipped;
+  const In *prologue_in = sample.in + start_x;
+
+  const uchar4 *aligned_in_char4 =
+      reinterpret_cast<const uchar4 *>(sample.in + start_x + bytes_skipped);
+
+  // prologue
+  for (int64_t idx = threadIdx.x; idx < bytes_skipped; idx += blockDim.x) {
+    prologue_tile[idx] = prologue_in[idx];
+  }
+  int64_t left_after_prologue = end_x - start_x - bytes_skipped;
+
+  // aligned load
+  for (int64_t idx = threadIdx.x; idx < left_after_prologue / 4; idx += blockDim.x) {
+    uchar4 in = aligned_in_char4[idx];
+    aligned_tile[idx * 4 + 0] = in.x;
+    aligned_tile[idx * 4 + 1] = in.y;
+    aligned_tile[idx * 4 + 2] = in.z;
+    aligned_tile[idx * 4 + 3] = in.w;
+  }
+  int64_t processed_in_main = (left_after_prologue / 4) * 4;
```

Comment:
One instruction instead of 3 (signed division by a power of 2 requires two instructions).
```suggestion
  int64_t processed_in_main = left_after_prologue & -4_i64;
```

Reply:
Done